File: app/permissions.py
"""Pure permission predicates shared by the FastAPI guards (dependencies/auth)
and the service layer's read-scoping.

These describe *read* visibility. The Admin Viewer privilege (is_admin_viewer)
grants admin-level read access to any user while leaving write/action
permissions governed solely by the base ``role`` — so these helpers are used
only to widen what a caller may see, never what they may change.

This module deliberately has no FastAPI dependency so services can import it
without pulling in web-layer code.
"""
from app.models.user import User, UserRole


def can_view_all(user: User) -> bool:
    """May read chair/admin-level surfaces (all submissions, reviews, decisions,
    files, presenters). True for program chairs, admins, and admin viewers."""
    return user.is_admin_viewer or user.role in (UserRole.PROGRAM_CHAIR, UserRole.ADMIN)


def can_view_admin(user: User) -> bool:
    """May read admin-only surfaces (users, audit log, conference settings,
    email/program templates). True for admins and admin viewers."""
    return user.is_admin_viewer or user.role == UserRole.ADMIN

File: frontend/lib/permissions.ts
// Client-side permission helpers, mirroring app/permissions.py on the backend.
//
// The Admin Viewer privilege (is_admin_viewer) grants admin-level *read* access
// to any user. It never grants the ability to act — write/action controls stay
// gated by the base role. So use the canView* helpers to decide what to *show*,
// and the canAct* helpers to decide what to *enable*.
import type { User } from "@/lib/api";

type U = Pick<User, "role" | "is_admin_viewer" | "is_reviewer"> | null | undefined;

// ── Acting (write) permissions — base role only ──────────────────────────────
export const canActAdmin = (u: U): boolean => u?.role === "admin";
export const canActChair = (u: U): boolean => u?.role === "admin" || u?.role === "program_chair";

// ── Viewing (read) permissions — base role OR the Admin Viewer flag ───────────
export const canViewAdmin = (u: U): boolean => !!u && (canActAdmin(u) || !!u.is_admin_viewer);
export const canViewAll = (u: U): boolean => !!u && (canActChair(u) || !!u.is_admin_viewer);

// True when the user is in the admin area purely as a read-only Admin Viewer for
// a given surface — i.e. they can see it but cannot act on it. Used to show the
// read-only banner and to disable controls.
export const isReadOnlyAdmin = (u: U): boolean => !canActChair(u) && !!u?.is_admin_viewer;

// May enter the /admin area at all (staff, reviewers, or admin viewers).
export const canEnterAdmin = (u: U): boolean =>
  !!u && (canViewAll(u) || u.role === "reviewer" || !!u.is_reviewer || !!u.is_admin_viewer);
